src/openms/source/ANALYSIS/ID/AScore.cpp

```diff
@@ -135,155 +153,157 @@ namespace OpenMS
     {
-      computeHighestPeptides(peptide_site_scores, highest_peptides, permutations);
-      phospho.setScore(peptideScore_(peptide_site_scores[highest_peptides[0].seq_1]));
-      //phospho.setSequence(AASequence(th_spectra[highest_peptides[0].seq_1].getName()));
+      determineHighestScoringPermutations(peptide_site_scores, phospho_sites, permutations);
+
+      // intialize score
+      phospho.setScore(peptideScore(peptide_site_scores[phospho_sites[0].seq_1]));
     }
+
     {
+      // initialize score with highest peptide score (aka heighest weighted score)
       multimap<double, Size> ranking;
-      for (Size it = 0; it < peptide_site_scores.size(); ++it)
+      for (Size i = 0; i != peptide_site_scores.size(); ++i)
       {
-        double current_score = peptideScore_(peptide_site_scores[it]);
-        ranking.insert(pair<double, Size>(current_score, it));
+        double weighted_score = peptideScore(peptide_site_scores[i]);
+        ranking.insert(pair<double, Size>(weighted_score, i));
       }
       phospho.setScore(ranking.rbegin()->first);
       phospho.setSequence(AASequence::fromString(th_spectra[ranking.rbegin()->second].getName()));
     }
+
     phospho.setCharge(hit.getCharge());
     phospho.setMetaValue("Search_engine_sequence", hit.getSequence().toString());
-    ///Calculate AScore
-    //Int add_ascores = (Int)highest_peptides.size()/permutations[0].size();
-    //Int count_ascores = 0;
-    //vector<ProbablePhosphoSites>::iterator winner = highest_peptides.begin();
-    //vector<ProbablePhosphoSites>::iterator actual = winner;
-    //double ascore_addition = 0.0;
-    //double highest_addition = 0.0;
+
     Int rank = 1;
-    for (vector<ProbablePhosphoSites>::iterator hp = highest_peptides.begin(); hp < highest_peptides.end(); ++hp)
+    for (vector<ProbablePhosphoSites>::iterator s_it = phospho_sites.begin(); s_it < phospho_sites.end(); ++s_it)
     {
       vector<RichPeakSpectrum> site_determining_ions;
-      compute_site_determining_ions(th_spectra, *hp, hit.getCharge(), site_determining_ions);
-      N = UInt(site_determining_ions[0].size());       // all possibilities have the same number
-      double p = (double)hp->peak_depth / 100;
-      Int n = 0;
-      //Auslagern
-      for (Size depth = 0; depth <  windows_with_all_peak_depths.size(); ++depth)       //each 100 m/z window
+      // Previously, the precursor charge was used here. This is clearly wrong and it is better to use charge 1 here.
+      computeSiteDeterminingIons(th_spectra, *s_it, 1, site_determining_ions);
+      Size N = site_determining_ions[0].size(); // all possibilities have the same number so take the first one
+      double p = static_cast<double>(s_it->peak_depth) / 100.0;
+
+      Size n_first = 0;
+      for (Size depth = 0; depth != windows_top10.size(); ++depth) // for each 100 m/z window
       {
-        n += numberOfMatchedIons_(site_determining_ions[0], windows_with_all_peak_depths[depth], (Size)p, fmt);
```

Comment:
bug: (Size)p
p is a probability and not the peak depth. Same bug occurs multiple times.